src/api/apiMethod.js

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright © 2017 Lisk Foundation
+ *
+ * See the LICENSE file at the top-level directory of this distribution
+ * for licensing information.
+ *
+ * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
+ * no part of this software, including this file, may be copied, modified,
+ * propagated, or distributed except according to the terms contained in the
+ * LICENSE file.
+ *
+ * Removal or modification of this copyright notice is prohibited.
+ *
+ */
+
+import { GET } from 'constants';
+import * as utils from './utils';
+
+// Bind to resource class
+const apiMethod = params => {
+	const spec = params || {};
+	const method = spec.method || GET;
+	const path = spec.path || '';
+	const urlParams = spec.urlParams || [];
+	const validator = spec.validator || null;
+	const defaultData = spec.defaultData || {};
+	const retry = spec.retry || false;
+
+	return function apiHandler(...args) {
+		// this refers to resource class
+		const self = this;
+		let fullURL = self.resourcePath + path;
```

Comment:
Any suggestion?
Also, why avoid let where `fullURL` can be changed afterwords?

Reply:
The question is: why make something that can change, rather than two things that don't change?

```js
// defined outside the function?
const addParamsToURL = (params, url, args) => {
    if (args.length < urlParams.length) {
        throw new Error('xxx');
    }
    const data = params.reduce((accumulator, param, i) => Object.assign({}, accumulator, { [param]: args[i] }), {});
    return utils.solveURLParams(url, data);
};

// inside the function
// ...
const urlWithoutParams = `${this.resourcePath}${path}`;
const fullURL = urlParams.length
    ? addParamsToURL(urlParams, urlWithoutParams, args)
    : urlWithoutParams;
```

Reply:
I came up with the idea without using `let` or two variables 👍